nix-tools/make-install-plan/MakeInstallPlan.hs

```diff
@@ -0,0 +1,124 @@
+      commandDescription = Nothing,
+      commandNotes = Nothing,
+      commandDefaultFlags = defaultNixStyleFlags (),
+      commandOptions = nixStyleOptions (const [])
+    }
+
+-- The following is adapted from cabal-install's Distribution.Client.CmdFreeze
+installPlanAction :: Verbosity -> ProjectConfig -> IO ()
+installPlanAction verbosity cliConfig = do
+  ProjectBaseContext {distDirLayout, cabalDirLayout, projectConfig, localPackages} <-
+    establishProjectBaseContext verbosity cliConfig OtherCommand
+
+  (_improvedPlan, elaboratedPlan, elaboratedSharedConfig, _tis, _at) <-
+    rebuildInstallPlan verbosity distDirLayout cabalDirLayout projectConfig localPackages
+
+  -- Write plan.json
+  Cabal.notice verbosity $ "Writing plan.json to " ++ distProjectCacheFile distDirLayout "plan.json"
+  writePlanExternalRepresentation distDirLayout elaboratedPlan elaboratedSharedConfig
+
+  -- Write cabal.freeze
+  let cabalFreezeFile = distProjectFile distDirLayout "freeze"
+  Cabal.notice verbosity $ "Wrote freeze file to " ++ cabalFreezeFile
+  writeProjectConfigFile cabalFreezeFile projectConfig
+
+  -- Write cabal files and their nix version
+  let cabalFilesDir = distDirectory distDirLayout </> "cabal-files"
+  Cabal.createDirectoryIfMissingVerbose verbosity True cabalFilesDir
+  Cabal.notice verbosity $ "Writing cabal files to " ++ cabalFilesDir
+
+  let ecps = [ecp | InstallPlan.Configured ecp <- InstallPlan.toList elaboratedPlan, not $ elabLocalToProject ecp]
+
+  for_ ecps $
+    \ElaboratedConfiguredPackage
+       { elabPkgSourceId,
+         elabPkgSourceLocation,
+         elabPkgSourceHash,
+         elabPkgDescriptionOverride
+       } -> do
+        let pkgFile = cabalFilesDir </> prettyShow (pkgName elabPkgSourceId) <.> "cabal"
+        let nixFile = cabalFilesDir </> prettyShow (pkgName elabPkgSourceId) <.> "nix"
+        for_ elabPkgDescriptionOverride $ \pkgTxt ->
+          do
+            -- raw
+            Cabal.info verbosity $ "Writing cabal file for " ++ prettyShow elabPkgSourceId ++ " to " ++ pkgFile
+            BSL.writeFile pkgFile pkgTxt
+            -- nix
+            let src = packageLocation2Src elabPkgSourceLocation elabPkgSourceHash
+            Cabal2Nix.cabal2nix False Cabal2Nix.MinimalDetails (Just src) (Cabal2Nix.OnDisk pkgFile)
+            >>= writeDoc nixFile . prettyNix
+
+packageLocation2Src :: PackageLocation local -> Maybe HashValue -> Cabal2Nix.Src
+packageLocation2Src pkgSrcLoc pkgSrcHash = case pkgSrcLoc of
+  LocalUnpackedPackage path -> Cabal2Nix.Path path
+  LocalTarballPackage path -> Cabal2Nix.Path path
+  RemoteTarballPackage uri _local -> Cabal2Nix.Repo (show uri) mSrcHash
+  RepoTarballPackage repo _packageId _local -> case repo of
+    (RepoLocalNoIndex lr _local) -> Cabal2Nix.Path (localRepoPath lr)
+    (RepoRemote rr _local) -> Cabal2Nix.Repo (show $ remoteRepoURI rr) mSrcHash
+    (RepoSecure rr _local) -> Cabal2Nix.Repo (show $ remoteRepoURI rr) mSrcHash
```

Comment:
I am a bit concerned that I had got these two urls wrong (I had accidentally included the package path into the repo url) and everything was still working correctly. Maybe it was just because the nix store was already populated? `pkgs.fetchurl` was called with the correct sha256 but with the wrong url.